### ISSUE TITLE
New Rule: `no-useless-override`

### ISSUE THREAD

**Issue by DetachHead** [eslint/eslint#15737]
### Rule details

same as `no-useless-constructor` but with overridden methods

### Related ECMAScript feature

how new? es6 classes?

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// fail
class Foo {
    bar() {}
}

class Bar extends Foo {
    bar() {
        super.bar()
    }
}


// pass
class Foo {
    bar() {}
}

class Bar extends Foo {
    bar() {
        console.log("doing something else")
        super.bar()
    }
}
```

### Why should this rule be in the core instead of a plugin?

because `no-useless-constructor` is already a core rule and i see no reason why it shouldn't also check for useless method overrides

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

originally raised here https://github.com/typescript-eslint/typescript-eslint/issues/4747

**Comment by ljharb** [comment]
This isn't the same at all.

Most subclasses live in a different file than their superclass, and most eslint rules do not work across files, so I think this is a nonstarter.

**Comment by bradzacher** [comment]
Why? If you are calling `super.foo()` from within a method `foo` and doing nothing else - isn't that going to be the same as not having the override at all?

I.e. it's the same as having `super()` within a constructor and nothing else.

Why do you need the parent class to understand this?

**Comment by ljharb** [comment]
That's a fair point, but in this case the example is defining a public field and not a prototype method - and it actually might matter that it's an own property (Object.assign/object spread, for example), because the parent class is likely to be a prototype property (because fields shouldn't have functions, generally).

**Comment by bradzacher** [comment]
Yeah there is definitely that difference between prototype member vs instance member with class methods vs class properties.

I.e. technically here the override here not 100% equivalent to not having it

```js
class Parent {
  method() { ... }
}

class Child extends Parent {
  method = () => { return super.parent() };
}
```

Though for a lint rule - I would postulate that this doesn't matter for the vast majority of code. As long as there isn't an autofixer - "false positiving" in the rare case would be safe.

**Comment by ljharb** [comment]
I suppose that's true. Although, I'm very skeptical this actually occurs almost ever - i would bet `super.foo()` is not a concept most people even know exists, and those who know of it are unlikely to make a method like this without doing something different.

**Comment by mdjermanovic** [comment]
`bar = () => { super.bar() }` does not have to be useless. Consider the following example:

```js
class A {
    x = 5;
    bar() {
        console.log(this.x);
    }
}

class B extends A {
    bar = () => {
        super.bar()
    }
}

class C extends A {
}

(0, new B().bar)(); // logs 5
(0, new C().bar)(); // throws TypeError: Cannot read property 'x' of undefined


```

**Comment by mdjermanovic** [comment]
> ```js
> class Foo {
>     bar = () => {}
> }
> 
> class Bar extends Foo {
>     bar = () => {
>         super.bar()
>     }
> }
> ```

This code actually has a bug - if you call `(new Bar()).bar()` it will throw an error because there is no `super.bar`. Arrow function instance fields are not methods.

**Comment by DetachHead** [comment]
oops, updated the code to use regular functions instead

**Comment by mdjermanovic** [comment]
I'm not sure if can ever consider `bar = () => { super.bar() }` as "useless". I believe it's either a bug when there's no `super.bar`, or a bit unusual way to bind it.

**Comment by ljharb** [comment]
```
class Bar extends Foo {
    bar() {
        super.bar()
    }
}
```
also isn't useless; it skips the return value and returns undefined.

**Comment by DetachHead** [comment]
yeah i pasted it from the original issue i raised on `typescript-eslint` where it would be able to determine the return type of `super.bar`, so i guess here it would need to only warn if you were returning it like so

```js
class Bar extends Foo {
    bar() {
        super.bar()
    }
}
```

**Comment by mdjermanovic** [comment]
@DetachHead the new proposal looks good to me, but since it doesn't relate to new ECMAScript features that reached stage 4 in the preceding 12 months, per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies) it doesn't qualify for inclusion in the core. You can always make a custom rule and publish a plugin.
